File: 0838/solution.py
class Solution:
    def pushDominoes(self, dominoes: str) -> str:
        res = list(dominoes)
        stack = []
        right_loc = float("-inf")
        for i, curr in enumerate(dominoes):
            if curr == ".":
                stack.append(i)
            elif curr == "R":
                while stack and right_loc != float("-inf"):
                    loc = stack.pop()
                    res[loc] = "R"
                stack = []
                right_loc = i
            else:
                while stack:
                    loc = stack.pop()
                    if right_loc == float("-inf"):
                        res[loc] = "L"
                    elif loc - right_loc < i - loc:
                        res[loc] = "R"
                    elif loc - right_loc > i - loc:
                        res[loc] = "L"
                right_loc = float("-inf")
        while stack and right_loc != float("-inf"):
            loc = stack.pop()
            res[loc] = "R"
        return "".join(res)

File: 0838/solution.ts
function pushDominoes(dominoes: string): string {
  let stack: Array<number> = [];
  const res = dominoes.split("");
  let rightLoc = -1;
  for (let i = 0; i < dominoes.length; i++) {
    const curr = dominoes[i];
    if (curr == ".") {
      stack.push(i);
    } else if (curr == "R") {
      if (rightLoc != -1) {
        while (stack.length) {
          const last = stack.pop()!;
          res[last] = "R";
        }
      } else {
        stack = [];
      }
      rightLoc = i;
    } else {
      while (stack.length) {
        const last = stack.pop()!;
        if (rightLoc == -1 || last - rightLoc > i - last) {
          res[last] = "L";
        } else if (last - rightLoc < i - last) {
          res[last] = "R";
        }
      }
      rightLoc = -1;
    }
  }
  while (rightLoc != -1 && stack.length) {
    const loc = stack.pop()!;
    res[loc] = "R";
  }
  return res.join("");
}
